packages/react-instantsearch-core/src/widgets/InstantSearch.tsx

```diff
@@ -136,7 +139,19 @@ class InstantSearch extends Component<Props, State> {
     nextProps: Props,
     prevState: State
   ): Partial<State> {
+    const nextIsControlled = isControlled(nextProps);
+    const previousSearchState = prevState.instantSearchManager.store.getState()
+      .widgets;
+    const nextSearchState = nextProps.searchState;
+
+    if (nextIsControlled && !isEqual(previousSearchState, nextSearchState)) {
```

Comment:
it's indeed more likely to cause issues with infinite rendering, than this being bad for perf